Not a valid gitRootDirectory on composite build project

Hi 

I have the structure of composite build project look likes

```
- workspace
    - app
        - .git 
        - build.gradle.kts 
    - lib
       - .git
       - build.gradle.kts
```

I added the appVersioning plugin to app/build.gradle.kts


```
appVersioning {
    fetchTagsWhenNoneExistsLocally.set(true)

    println("My root folder is: ${rootProject.file("./")}")

    overrideVersionName { gitTag, _, _ ->
        // TODO generate a String from the given gitTag, providers, build variant
        "${gitTag.rawTagName}-${gitTag.commitHash}"
    }
}
```

but I encountered the error:

```
> Configure project :app
My root folder is: /Users/ken/Developer/projects/2021/workspace/app


> Task :app:generateAppVersionInfoForDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:generateAppVersionInfoForDebug'.
> A failure occurred while executing io.github.reactivecircus.appversioning.tasks.GenerateAppVersionInfoWorkAction
   > Android App Versioning Gradle Plugin works with git tags but root project 'browser-app' is not a git root directory, and a valid gitRootDirectory is not provided.

```


I confirmed that the app folder contains the.git directory


Thank you

Hi @ychescale9 

Thanks for your reply, 

I added the line to the build.gradle.kts

```
appVersioning {
    fetchTagsWhenNoneExistsLocally.set(true)

    println("My root folder is: ${rootProject.file("app/")}")

    gitRootDirectory.set(rootProject.file("app/"))

    ......
}
```

and the error still exists, I think the root directory is wrong after the line is added.


```
BUILD FAILED in 3s
1 actionable task: 1 executed
❯ ./gradlew generateAppVersionInfoForDebug

> Configure project :app
My root folder is: /Users/ken/Developer/projects/2021/workspace/app/app

> Task :app:generateAppVersionInfoForDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:generateAppVersionInfoForDebug'.
> A failure occurred while executing io.github.reactivecircus.appversioning.tasks.GenerateAppVersionInfoWorkAction
   > Android App Versioning Gradle Plugin works with git tags but root project 'browser-app' is not a git root directory, and a valid gitRootDirectory is not provided.


```

--- a/README.md
+++ b/README.md
@@ -411,6 +411,21 @@ appVersioning {
 }
 ```
 
+### Bare git repository
+
+If your `.git` is a symbolic link to a **bare** git repository, you need to explicitly specify the directory of the bare git repository:
+
+```kt
+appVersioning {
+    /**
+     * Bare Git repository directory.
+     * Use this to explicitly set the directory of a bare git repository (e.g. `app.git`) instead of the standard `.git`.
+     * Setting this will override the value of [gitRootDirectory] property.
+     */
+    bareGitRepoDirectory.set(rootProject.file("../.repo/projects/app.git")) // if the .git directory in the Gradle project root is a symlink to app.git.
+}
+```
+
 ## App versioning on CI
 
 For performance reason many CI providers only fetch a single commit by default when checking out the repository. For **app-versioning** to work we need to make sure Git tags are also fetched. Here's an example for doing this with [GutHub Actions](https://github.com/actions/checkout):

--- a/api/app-versioning.api
+++ b/api/app-versioning.api
@@ -1,5 +1,6 @@
 public class io/github/reactivecircus/appversioning/AppVersioningExtension {
 	public static final field Companion Lio/github/reactivecircus/appversioning/AppVersioningExtension$Companion;
+	public final fun getBareGitRepoDirectory ()Lorg/gradle/api/file/DirectoryProperty;
 	public final fun getEnabled ()Lorg/gradle/api/provider/Property;
 	public final fun getFetchTagsWhenNoneExistsLocally ()Lorg/gradle/api/provider/Property;
 	public final fun getGitRootDirectory ()Lorg/gradle/api/file/DirectoryProperty;

--- a/src/main/kotlin/io/github/reactivecircus/appversioning/AppVersioningExtension.kt
+++ b/src/main/kotlin/io/github/reactivecircus/appversioning/AppVersioningExtension.kt
@@ -41,6 +41,13 @@ open class AppVersioningExtension internal constructor(objects: ObjectFactory) {
     val gitRootDirectory = objects.directoryProperty().convention(null)
 
     /**
+     * Bare Git repository directory.
+     * Use this to explicitly set the directory of a bare git repository (e.g. `app.git`) instead of the standard `.git`.
+     * Setting this will override the value of [gitRootDirectory] property.
+     */
+    val bareGitRepoDirectory = objects.directoryProperty().convention(null)
+
+    /**
      * Custom glob pattern for matching git tags.
      */
     val tagFilter = objects.property<String>().convention(null)

--- a/src/main/kotlin/io/github/reactivecircus/appversioning/AppVersioningPlugin.kt
+++ b/src/main/kotlin/io/github/reactivecircus/appversioning/AppVersioningPlugin.kt
@@ -14,6 +14,7 @@ import org.gradle.api.tasks.TaskProvider
 import org.gradle.kotlin.dsl.getByType
 import org.gradle.kotlin.dsl.withType
 import org.gradle.language.nativeplatform.internal.BuildType
+import java.io.File
 import java.util.concurrent.atomic.AtomicBoolean
 
 /**
@@ -68,21 +69,14 @@ class AppVersioningPlugin : Plugin<Project> {
 
     private fun Project.registerGenerateAppVersionInfoTask(
         variant: ApplicationVariant,
-        extension: AppVersioningExtension
+        extension: AppVersioningExtension,
     ): TaskProvider<GenerateAppVersionInfo> = tasks.register(
         "${GenerateAppVersionInfo.TASK_NAME_PREFIX}For${variant.name.capitalize()}",
         GenerateAppVersionInfo::class.java
     ) {
         group = APP_VERSIONING_TASK_GROUP
         description = "${GenerateAppVersionInfo.TASK_DESCRIPTION_PREFIX} for the ${variant.name} variant."
-
-        gitRefsDirectory.set(project.rootProject.file(GIT_REFS_DIRECTORY).let { rootGradleProject ->
-            if (rootGradleProject.exists()) rootGradleProject else {
-                extension.gitRootDirectory.takeIf { it.isPresent }?.let { gitRootDirectory ->
-                    gitRootDirectory.asFile.orNull?.resolve(GIT_REFS_DIRECTORY)?.takeIf { it.exists() }
-                }
-            }
-        })
+        gitRefsDirectory.set(findGitRefsDirectory(extension))
         rootProjectDirectory.set(project.rootProject.rootDir)
         rootProjectDisplayName.set(project.rootProject.displayName)
         fetchTagsWhenNoneExistsLocally.set(extension.fetchTagsWhenNoneExistsLocally)
@@ -123,14 +117,31 @@ class AppVersioningPlugin : Plugin<Project> {
         buildVariantName.set(variantName)
     }
 
+    private fun Project.findGitRefsDirectory(extension: AppVersioningExtension): File? {
+        return when {
+            extension.bareGitRepoDirectory.isPresent -> {
+                extension.bareGitRepoDirectory.let { bareGitRepoDirectory ->
+                    bareGitRepoDirectory.asFile.orNull?.resolve(REFS_DIRECTORY)?.takeIf { it.exists() }
+                }
+            }
+            extension.gitRootDirectory.isPresent -> {
+                extension.gitRootDirectory.let { gitRootDirectory ->
+                    gitRootDirectory.asFile.orNull?.resolve(STANDARD_GIT_REFS_DIRECTORY)?.takeIf { it.exists() }
+                }
+            }
+            else -> project.rootProject.file(STANDARD_GIT_REFS_DIRECTORY).takeIf { it.exists() }
+        }
+    }
+
     companion object {
         private const val MIN_GRADLE_VERSION = "6.8"
         private const val MIN_AGP_VERSION = "7.0.0-beta04"
     }
 }
 
-internal const val APP_VERSIONING_TASK_GROUP = "versioning"
-internal const val APP_VERSIONING_TASK_OUTPUT_DIR = "outputs/app_versioning"
-internal const val GIT_REFS_DIRECTORY = ".git/refs"
-internal const val VERSION_CODE_RESULT_FILE = "version_code.txt"
-internal const val VERSION_NAME_RESULT_FILE = "version_name.txt"
+private const val APP_VERSIONING_TASK_GROUP = "versioning"
+private const val APP_VERSIONING_TASK_OUTPUT_DIR = "outputs/app_versioning"
+private const val REFS_DIRECTORY = "refs"
+private const val STANDARD_GIT_REFS_DIRECTORY = ".git/$REFS_DIRECTORY"
+private const val VERSION_CODE_RESULT_FILE = "version_code.txt"
+private const val VERSION_NAME_RESULT_FILE = "version_name.txt"

